### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by jeehoonkang**
Now crossbeam-epoch's test coverage is largely the same with coco.

A caveat is that we need to do `RUST_TEST_THREADS=1 cargo test` instead of `cargo test` for sequentially executing test codes.  This is because two tests (`incremental` and `buffering`) are correct under the assumption that the thread in concern is the only thread that interacts with the global garbage queue.  I think the best way to deal with this problem is using custom instances (realms) of EBR instead of the global instance in test codes.

I'd like to suggest we publish crossbeam-epoch 0.1.0 after merging this, #16, and the custom realms proposal discussed in https://github.com/crossbeam-rs/rfcs/issues/12.  I'd like to see how existing projects will adopt the new Crossbeam™.  Cf. #14.

### CODE REVIEW

**Comment by ghost** [comment]
Regarding parallel test execution, check out [this](https://github.com/rust-lang/rust/issues/33519#issuecomment-314224553) comment. Also, it seems `RUST_TEST_THREADS=1` is [deprecated](https://stackoverflow.com/questions/44947914/how-to-limit-the-number-of-test-threads-in-cargo-toml).

I'd prefer to avoid relying on the flag for correct test execution. I think `cargo test` should *always* pass without requiring any additional flags. The recommended workaround to make tests run sequentially would be to create a mutex and lock it just before running a test.

Also, note that if you write tests in `tests/foo1.rs` and `tests/foo2.rs`, these two files will be tested one after another (but tests within each file will be run in parallel). However, all tests in `src/**` will be run concurrently.

Other than that, this looks pretty good. :) Did you port all relevant tests from coco now? Are there any we're still missing?

**Comment by jeehoonkang** [comment]
- In my opinion, the test should really be run in parallel.  (After all, Crossbeam is for concurrency and parallelism!)  I think this PR can wait for the custom realm proposal to be landed, after which test code can use their own custom realm, no longer requiring sequential execution of test codes.

- I ported every test except for `garbage.rs`'s `smoke()` and `flush_pending()`.  They are testing pending bags in coco, which have no corresponding notion in the current implementation in `crossbeam-epoch`.

**Comment by jeehoonkang** [comment]
Hi!  this is rebased on #21 (the `Collector` API), and thanks to it, now we can safely run tests in parallel :)  It is quite comprehensive test cases for `Collector`, which are mostly cut-and-pasted from coco.  I think it's ready to be merged.

File: src/sync/list.rs [hunk 1]
```diff
@@ -93,17 +96,21 @@ impl<T> List<T> {
     }
 
     /// Inserts `data` into the head of the list.
+    #[inline]
     pub fn insert<'scope>(&'scope self, data: T, scope: &'scope Scope) -> Ptr<'scope, Node<T>> {
         Self::insert_internal(&self.head, data, scope)
     }
 
-    /// Inserts `data` after `to`.
+    /// Inserts `data` after into the head of the list.
```
Comment (Vtec234) [inline src/sync/list.rs]: Should this be "inserts `data` after `after` into the list."?

**Review by Vtec234 — APPROVED** [review]
Something funky is going on with the GitHub diff display after the rebase, but I checked it locally and it looks fine.

**Review by arthurprs — APPROVED** [review]
Awesome work on this refactor/rfc :tada:

File: src/sync/list.rs (hunk 1 above)
Comment (jeehoonkang) [reply src/sync/list.rs]: Thanks!  I just pushed a new commit that changes the doc as you said.

File: src/collector.rs [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        static DESTROYS: AtomicUsize = ATOMIC_USIZE_INIT;
+
+        let collector = Collector::new();
+        let handle = collector.handle();
+
+        handle.pin(|scope| unsafe {
+            for _ in 0..COUNT {
+                let a = Owned::new(7i32).into_ptr(scope).as_raw() as usize;
+                scope.defer(move || {
+                    drop(Owned::from_raw(a as *mut i32));
```
Comment (ghost) [inline src/collector.rs]: I think we can use `a.into_owned()` here, and remove the `.as_raw() as usize` part above.

File: src/collector.rs [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
```
Comment (ghost) [inline src/collector.rs]: Same here.

**Review by ghost — APPROVED** [review]
Just a few nits.

**Comment by jeehoonkang** [comment]
@stjepang Thanks!  I just changed the code as you said.

**Pull request merged**